server/sonar-server/src/main/java/org/sonar/server/computation/component/PathAwareVisitorAdapter.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * SonarQube, open source software quality management tool.
+ * Copyright (C) 2008-2014 SonarSource
+ * mailto:contact AT sonarsource DOT com
+ *
+ * SonarQube is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.computation.component;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A adapter of the {@link PathAwareVisitor} to be able to visit only some component types
+ */
+public abstract class PathAwareVisitorAdapter<T> implements PathAwareVisitor<T>, Visitor {
+  private final Component.Type maxDepth;
+  private final Order order;
+  private final StackElementFactory<T> factory;
+  private final DequeBasedPathAdapter<T> stack = new DequeBasedPathAdapter<>();
+
+  public PathAwareVisitorAdapter(Component.Type maxDepth, Order order, StackElementFactory<T> factory) {
+    this.maxDepth = requireNonNull(maxDepth);
+    this.order = requireNonNull(order);
+    this.factory = requireNonNull(factory, "Factory can not be null");
+  }
+
+  public DequeBasedPath<T> getStack() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the "@Override" annotation above this method signature [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)